---
title: "Stone et al. 2023 methylation analysis"
author: "Carl Stone, Gwyneth Boyer, & Megan Behringer"
output:
  github_document:
    df_print: paged
---

# CoMMA: Comparison of Microbial Methylated Adenines

Below is the code for replicating all analyses in Stone et al. 2023. The functions sourced in block 3 can also be found in this repository. CoMMA can be found at https://github.com/carl-stone/CoMMA. More details on general nanopore methylation analysis in bacteria can be found in the CoMMA repository.

## Raw data and pre-processing

```{r intall_required_packages, eval=FALSE, include=FALSE}
BiocManager::install(c("tidyverse",
                       "methylKit",
                       "cowplot",
                       "ggrepel",
                       "vegan",
                       "ggfortify",
                       "moments",
                       "RColorBrewer",
                       "R.utils",
                       "boot",
                       "preprocessCore",
                       "rstatix",
                       "modelr",
                       "splines",
                       "MASS",
                       "ggpubr",
                       "circlize",
                       "eulerr",
                       "ragg",
                       "Metrics",
                       "quantreg",
                       "remotes",
                       "gtools",
                       "readxl",
                       "ape",
                       "ggupset",
                       "clusterProfiler",
                       "AnnotationHub",
                       "org.EcK12.eg.db",
                       "DOSE"))
install_github("pmartinezarbizu/pairwiseAdonis/pairwiseAdonis")
install_github("carl-stone/CoMMA")

```

```{r load_required_packages, message=FALSE, warning=FALSE}
library(tidyverse)
library(methylKit)
library(cowplot)
library(ggrepel)
library(vegan)
library(ggfortify)
library(moments)
library(RColorBrewer)
library(R.utils)
library(boot)
library(preprocessCore)
library(rstatix)
library(modelr)
library(splines)
library(MASS)
library(ggpubr)
library(circlize)
library(eulerr)
library(ragg)
library(Metrics)
library(quantreg)
library(remotes)
library(gtools)
library(readxl)
library(circlize)
library(ggupset)
library(pairwiseAdonis)
library(enrichplot)
library(DOSE)
library(clusterProfiler)
library(org.EcK12.eg.db)

if (!require("ggview")) {
  remotes::install_github("idmn/ggview")
  library(ggview)
}
```

```{r raw_data_input, warning=FALSE}
# Set wd to Behringer_Lab_Box_Drive
if (Sys.info()['nodename'] == "AS707CQ6X2" |
    Sys.info()['nodename'] == "carls-mbp.lan" |
    Sys.info()['nodename'] == "Carls-MacBook-Pro.local") {
  setwd("/Users/carlstone/Library/CloudStorage/Box-Box/Behringer_Lab_Box_Drive")
} else if (Sys.info()['nodename'] == "DESKTOP-QM8UFU1") {
  setwd("C:/Users/carls/Box/Behringer_Lab_Box_Drive")
} else {
  print('Cannot set path on this device, do it yourself.')
}
source('Projects/LongTermExpEvo/MethylationProject/Rscripts/functions.R')

allCallers_WT <-  read.table(file = "Projects/LongTermExpEvo/MethylationProject/Benchmarking/WT_All_Callers_m6A.txt",
                             sep = "\t", 
                             header = TRUE)

euler_data <- read_excel(path = 'Projects/LongTermExpEvo/MethylationProject/Benchmarking/WT_allCallers_annotated.xlsx')

mgAllSamples <- as_tibble(read.table(file = 'Projects/LongTermExpEvo/MethylationProject/ComparativeAnalysis/RawDataFiles/SortedFiles/MethylPercent_Coverage_All_NoBlank.txt',
                                     sep = '\t',
                                     header = T))

genome_sites <- read_tsv(file = 'Projects/LongTermExpEvo/MethylationProject/Ecoli_Annotation/all_site_annotations.txt')

siteMutations <- read_tsv(file = '/Users/carlstone/Library/CloudStorage/Box-Box/Behringer_Lab_Box_Drive/Projects/LongTermExpEvo/MethylationProject/GATC_Mutation_Data/SummaryData/GATCsiteSummary.txt')

```

```{r raw_data_processing}

allCallers_WT <- allCallers_WT %>% 
  dplyr::select(c(DS_Position, DS_Strand_WT1, 
                  contains("Coverage"), contains("Methyl"))) %>% 
  dplyr::select(!contains("Var")) %>% 
  dplyr::rename(
    Position = DS_Position,
    Strand = DS_Strand_WT1,
  ) %>% 
  dplyr::rename_with(~ gsub("Methyl_Percent", "MethylPercent", .x)) %>% 
  drop_na()

allCallers_WT <- allCallers_WT %>% 
  filter(DS_Coverage_WT1 > 10 &
         DS_Coverage_WT2 > 10 &
         DS_Coverage_WT3 > 10) %>% 
  filter(mC_Coverage_WT1 > 10 &
         mC_Coverage_WT2 > 10 &
         mC_Coverage_WT3 > 10) %>% 
  filter(Mg_Coverage_WT1 > 10 &
         Mg_Coverage_WT2 > 10 &
         Mg_Coverage_WT3 > 10) %>% 
  filter(Tm_Coverage_WT1 > 10 &
         Tm_Coverage_WT2 > 10 &
         Tm_Coverage_WT3 > 10)

# Add mean methylation for each site
allCallers_WT$DS_MethylPercent_Mean <- rowMeans(cbind(allCallers_WT$DS_MethylPercent_WT1,
                                                      allCallers_WT$DS_MethylPercent_WT2,
                                                      allCallers_WT$DS_MethylPercent_WT3))
allCallers_WT$mC_MethylPercent_Mean <- rowMeans(cbind(allCallers_WT$mC_MethylPercent_WT1,
                                                      allCallers_WT$mC_MethylPercent_WT2,
                                                      allCallers_WT$mC_MethylPercent_WT3))
allCallers_WT$Mg_MethylPercent_Mean <- rowMeans(cbind(allCallers_WT$Mg_MethylPercent_WT1,
                                                      allCallers_WT$Mg_MethylPercent_WT2,
                                                      allCallers_WT$Mg_MethylPercent_WT3))
allCallers_WT$Tm_MethylPercent_Mean <- rowMeans(cbind(allCallers_WT$Tm_MethylPercent_WT1,
                                                      allCallers_WT$Tm_MethylPercent_WT2,
                                                      allCallers_WT$Tm_MethylPercent_WT3))

allCallers_WT_long <- allCallers_WT %>% 
  pivot_longer(
    cols = DS_Coverage_WT1:Tm_MethylPercent_WT3,
    names_to = c("Method", ".value", "Sample"),
    names_sep = "_"
  ) 

allCallers_WT_average_long <- allCallers_WT %>% 
  dplyr::select(c(Position, Strand, contains("Mean"))) %>% 
  pivot_longer(
    cols = contains("Mean"),
    names_to = c("Method", ".value", "Sample"),
    names_sep = "_"
  ) 

mgAllSamples <- mgAllSamples %>%
  dplyr::mutate(across(c(cov_WT1:methyl_129.3), as.numeric))
mgAllSamples$Position <- mgAllSamples$Position + 1
mgAllSamplesLong <- mgAllSamples %>%
  pivot_longer(
    cols = cov_WT1:methyl_129.3,
    names_to = c(".value", "sample"),
    names_sep = "_"
  ) %>%
  filter(cov >= 10) %>% 
  drop_na()
mgAllSamples <- mgAllSamplesLong %>% 
  pivot_wider(names_from = sample,
              values_from = c(cov, methyl)) %>% 
  drop_na()
mgAllSamplesLong$genotype <- mgAllSamplesLong$sample
mgAllSamplesLong <- mgAllSamplesLong %>% 
  mutate(genotype = dplyr::recode(
    genotype,
    WT1 = 'anc',
    WT2 = 'anc',
    WT3 = 'anc',
    '125.6' = 'WT',
    '125.7' = 'WT',
    '129.1' = 'WT',
    '129.3' = 'WT',
    '113.1' = 'MMR',
    '113.2' = 'MMR',
    '126.2' = 'MMR',
    '126.4' = 'MMR'
  ))

mgAllSamplesLong <- mgAllSamplesLong %>% 
  dplyr::rename(beta = "methyl")

# Set alpha to 1 over the largest coverage value 
mgAllSamplesLong$m <- methylBtoM(mgAllSamplesLong$beta, alpha = 0.01)

mgAllSamplesLong$genotype <- mgAllSamplesLong$sample
mgAllSamplesLong <- mgAllSamplesLong %>% 
  mutate(genotype = dplyr::recode(
    genotype,
    WT1 = 'anc',
    WT2 = 'anc',
    WT3 = 'anc',
    '125.6' = 'WT',
    '125.7' = 'WT',
    '129.1' = 'WT',
    '129.3' = 'WT',
    '113.1' = 'MMR',
    '113.2' = 'MMR',
    '126.2' = 'MMR',
    '126.4' = 'MMR'
  ))

WT_average <- mgAllSamplesLong %>% 
  filter(sample == 'WT1' |
           sample == 'WT2' |
           sample == 'WT3') %>% 
  group_by(Position) %>% 
  dplyr::summarize(cov = mean(cov),
            beta = mean(beta),
            m = mean(m))
WT_average <- left_join(WT_average,
                        allCallers_WT[, 1:2],
                        by = "Position")
WT_average <- WT_average %>%
  mutate(Strand = str_replace(Strand, "Plus", "+")) %>% 
  mutate(Strand = str_replace(Strand, "Minus", "-"))
```

## Methylation caller benchmarking

```{r methyl_distributions_by_caller}

methyl_distribution_by_caller <- allCallers_WT_average_long %>%
  ggplot(aes(x = MethylPercent*100, fill = Method)) +
  geom_density(alpha = 0.85) +
  scale_fill_brewer(palette = "Pastel1") +
  scale_y_continuous(expand = expansion(mult = c(0, 0.05))) +
  scale_x_continuous(expand = expansion(mult = c(0,0))) +
  coord_cartesian(xlim = c(70, 100)) +
  theme_stone() +
  theme(legend.position = c(0.25, 0.7),
        legend.box.background = element_rect(color = "black",
                                             linewidth = 1)) +
  labs(x = "Percent methylated reads",
       y = "Density",
       fill = paste0("Methylation\ncaller"))

methyl_distribution_by_caller
```

```{r euler_plots}
# Euler diagram showing overlap between replicates of each caller ---------
# and between consistent sites of all four callers


euler_data <- euler_data %>% 
  dplyr::select(Position, DS_60_1:Tm_count)

pos <- euler_data$Position

# Can flip around A and B to find intersect of methylated sites (1 = A) or 
# intersect of unmethylated sites (1 = B)
DS_sites <- euler_data %>% 
  dplyr::select(DS_60_1:DS_60_3) %>% 
  mutate(across(everything(), ~replace(., . == 1, 'A'))) %>% 
  mutate(across(everything(), ~replace(., . == 0, 'B')))
DS_sites <- as.matrix(DS_sites)
rownames(DS_sites) <- pos
colnames(DS_sites) <- c('Rep 1','Rep 2','Rep 3')
mC_sites <- euler_data %>% 
  dplyr::select(mC_60_1:mC_60_3) %>% 
  mutate(across(everything(), ~replace(., . == 1, 'A'))) %>% 
  mutate(across(everything(), ~replace(., . == 0, 'B')))
mC_sites <- as.matrix(mC_sites)
rownames(mC_sites) <- pos
colnames(mC_sites) <- c('Rep 1','Rep 2','Rep 3')
Mg_sites <- euler_data %>% 
  dplyr::select(Mg_60_1:Mg_60_3) %>% 
  mutate(across(everything(), ~replace(., . == 1, 'A'))) %>% 
  mutate(across(everything(), ~replace(., . == 0, 'B')))
Mg_sites <- as.matrix(Mg_sites)
rownames(Mg_sites) <- pos
colnames(Mg_sites) <- c('Rep 1','Rep 2','Rep 3')
Tm_sites <- euler_data %>% 
  dplyr::select(Tm_60_1:Tm_60_3) %>% 
  mutate(across(everything(), ~replace(., . == 1, 'A'))) %>% 
  mutate(across(everything(), ~replace(., . == 0, 'B')))
Tm_sites <- as.matrix(Tm_sites)
rownames(Tm_sites) <- pos
colnames(Tm_sites) <- c('Rep 1','Rep 2','Rep 3')

DS_euler <- euler(DS_sites, shape = 'ellipse')
mC_euler <- euler(mC_sites, shape = 'ellipse')
Mg_euler <- euler(Mg_sites, shape = 'ellipse')
Tm_euler <- euler(Tm_sites, shape = 'ellipse')
DS_reps_euler_plot <- plot(DS_euler, quantities = TRUE)
mC_reps_euler_plot <- plot(mC_euler, quantities = TRUE)
Mg_reps_euler_plot <- plot(Mg_euler, quantities = TRUE)
Tm_reps_euler_plot <- plot(Tm_euler, quantities = TRUE)
all_callers_all_replicates_grid <- plot_grid(DS_reps_euler_plot,
                                             mC_reps_euler_plot,
                                             Mg_reps_euler_plot,
                                             Tm_reps_euler_plot,
                                             labels = c("DeepSignal",
                                                        "mCaller",
                                                        "Megalodon",
                                                        "Tombo"))

DS_under <- euler_data %>% 
  filter(DS_count == 0) %>% 
  dplyr::select(Position)
mC_under <- euler_data %>% 
  filter(mC_count == 0) %>% 
  dplyr::select(Position)
Mg_under <- euler_data %>% 
  filter(Mg_count == 0) %>% 
  dplyr::select(Position)
Tm_under <- euler_data %>% 
  filter(Tm_count == 0) %>% 
  dplyr::select(Position)

# make a list of all consistently undermethylated sites for each caller
all_undermethylated_sites <- list(DeepSignal = DS_under$Position,
                                  mCaller = mC_under$Position,
                                  Megalodon = Mg_under$Position,
                                  Tombo = Tm_under$Position)

all_callers_euler <- euler(all_undermethylated_sites, shape = 'ellipse')
all_callers_euler_plot <- plot(all_callers_euler,
                               quantities = TRUE,
                               fills = brewer.pal(4, "Pastel1"),
                               adjust_labels = TRUE)

#ggsave(filename = paste0('euler_undermethylated_sites_', Sys.Date(), ".png") ,
#       plot = all_callers_euler_plot,
#       path = 'Projects/LongTermExpEvo/MethylationProject/Manuscript/Figures',
#       device = agg_png,
#       width = 8,
#       height = 6.5
#)

```


```{r inconsistently_called_sites}

# Inconsistently called sites bar graph
n_inconsistent_sites <- tibble(Caller = c("DeepSignal","mCaller","Megalodon","Tombo"),
                               nSites = c(2340, 3577, 162, 294))
n_sites_total <- 38236
n_inconsistent_sites$psites <- c(2340/n_sites_total*100, 3577/n_sites_total*100, 162/n_sites_total*100, 294/n_sites_total*100)
n_inconsistent_sites$Caller <- forcats::as_factor(n_inconsistent_sites$Caller)

plot_inconsistent_sites <- ggplot(n_inconsistent_sites, aes(x = Caller,
                                                            y = psites, 
                                                            fill = Caller)) +
  geom_bar(stat = "identity", color = 'black') +
  scale_fill_brewer(palette = "Pastel1") +
  scale_y_continuous(limits = c(0,10),
                     expand = expansion(mult = c(0,0.05)),
                     breaks = seq(0,10, 2),
                     minor_breaks = seq(1:10)) +
  theme_stone() +
  theme(legend.position = 'none') +
  labs(x = 'Methylation caller program',
       y = "Percent sites inconsistently classified")

plot_inconsistent_sites
```

```{r known_undermethylated_sites}
known_nonmethylated_sites <- tibble(Strand = c('+','-','+','-'),
                                    Locus = c('srlR','srlR','mtlA','mtlA'),
                                    DeepSignal = c(0.35, 0.70, 0.30, 0.62),
                                    mCaller = c(0.16, 0.18, 0.44, 0.24),
                                    Megalodon = c(0.37, 0.15, 0.25, 0.47),
                                    Tombo = c(0.12, 0.38, 0.80, 0.31)) %>% 
  pivot_longer(cols = DeepSignal:Tombo,
               names_to = 'Caller',
               values_to = 'pMethylated')

plot_known_nonmethylated_sites <- ggplot(known_nonmethylated_sites,
                                         aes(x = Strand, y = pMethylated*100, fill = Caller)) +
  geom_bar(stat = 'identity', position = position_dodge(), color = 'black') +
  facet_wrap(~Locus) +
  geom_hline(yintercept = 60, color = 'red') +
  scale_fill_brewer(palette = "Pastel1") +
  scale_y_continuous(expand = expansion(mult = c(0, 0.05))) +
  theme_stone() +
  theme(axis.text.x = element_text(size = 14),
        legend.position = "none",
        strip.text.x = element_text(face = "bold",
                                    size = 12),
        strip.background = element_rect(fill = "white")) +
  labs(y = 'Percent methylated reads')

plot_known_nonmethylated_sites
```

```{r figure_2}

figure_2 <- plot_grid(methyl_distribution_by_caller,
                      all_callers_euler_plot,
                      plot_inconsistent_sites,
                      plot_known_nonmethylated_sites,
                      labels = c("A", "B", "C", "D"))

ggview(figure_2, height = 7, width = 8)

#ggsave(figure_2,
#       filename = "/Users/carlstone/Library/CloudStorage/Box-Box/Behringer_Lab_Box_Drive/Projects/LongTermExpEvo/MethylationProject/Manuscript/Figures/figure_2.pdf",
#       width = 8,
#       height = 7)

#ggsave(figure_2,
#       filename = "/Users/carlstone/Library/CloudStorage/Box-Box/Behringer_Lab_Box_Drive/Projects/LongTermExpEvo/MethylationProject/Manuscript/Figures/figure_2.png",
#       width = 8,
#       height = 7,
#       dpi = 320)

```


## WT methylome characterization

```{r annotate_WT}
WT_average_annotated <- annotateMethylSites(methyl_df = WT_average,
                                             meta_df = genome_sites,
                                             location = 'Position')
WT_average_annotated_long <-  WT_average_annotated %>% 
  dplyr::select(!'Transcription-Units') %>% 
  pivot_longer(cols = Genes:'Origin-of-Replication',
               names_to = 'feature_type',
               values_to = 'feature_name') %>% 
  drop_na()
```

```{r WT_dist}
WT_dist <- mgAllSamplesLong %>% 
  filter(sample == 'WT1' |
           sample == 'WT2' |
           sample == 'WT3') %>% 
  ggplot(aes(x = m,
             color = sample)) +
  geom_density() +
  theme_stone() +
  theme(legend.position = 'none') + 
  scale_color_brewer(palette = 'Dark2')
WT_ecdf <- mgAllSamplesLong %>% 
  filter(sample == 'WT1' |
           sample == 'WT2' |
           sample == 'WT3') %>% 
  ggplot(aes(x = m,
             color = sample)) +
  stat_ecdf() +
  theme_stone() +
  scale_color_brewer(palette = 'Dark2')
WT_distributions <- plot_grid(WT_dist, WT_ecdf,
                              rel_widths = c(0.75, 1))

# mgAllSamples_normalized_long %>%
#   filter(sample == 'WT1' |
#            sample == 'WT2' |
#            sample == 'WT3') %>%
#   {pairwise_ks_test(value = .$m,
#                      group = .$sample)}
# mgAllSamples_normalized_long %>%
#   filter(sample == 'WT1' |
#            sample == 'WT2' |
#            sample == 'WT3') %>%
#   kruskal_test(m ~ sample)
# mgAllSamples_normalized_long %>%
#   filter(sample == 'WT1' |
#            sample == 'WT2' |
#            sample == 'WT3') %>%
#   dunn_test(m ~ sample)
```


```{r beta_x_position_S1}
beta_x_position <- WT_average_annotated_long %>%
  group_by(Position) %>% 
  ggplot(aes(x = Position, y = beta*100)) + 
  geom_point(size = 1) +
  theme_stone() +
  scale_x_continuous(expand = expansion(mult = c(0.01, 0.01))) +
  geom_hline(yintercept = median(WT_average_annotated_long$beta*100),
             color = "red") +
  labs(x = "Genome position",
       y = "Percent methylated reads")

beta_x_position

ggview(beta_x_position,
       width = 8,
       height = 4)

#ggsave(plot = beta_x_position,
#       filename = "/Users/carlstone/Library/CloudStorage/Box-Box/Behringer_Lab_Box_Drive/Projects/LongTermExpEvo/MethylationProject/Manuscript/Figures/figure_S1.png",
#       width = 8,
#       height = 4,
#       dpi = 320)
#
#ggsave(plot = beta_x_position,
#       filename = "/Users/carlstone/Library/CloudStorage/Box-Box/Behringer_Lab_Box_Drive/Projects/LongTermExpEvo/MethylationProject/Manuscript/Figures/figure_S1.pdf",
#       width = 8,
#       height = 4)
```


```{r anc_cov_distributions_strand_S2}
figure_s2 <- WT_average %>% 
  drop_na() %>% 
  ggplot(aes(x = cov, fill = Strand)) +
  geom_density(alpha = 0.75) +
  theme_stone() +
  scale_fill_brewer(palette = "Pastel1") +
  labs(x = "Read coverage",
       y = "Density") +
  geom_vline(xintercept = 57,
             linetype = "dashed") +
  scale_y_continuous(expand = expansion(mult = c(0.001, 0.05)))

ggview(figure_s2,
       width = 6,
       height = 4)

#ggsave(figure_s2,
#       filename = "/Users/carlstone/Library/CloudStorage/Box-Box/Behringer_Lab_Box_Drive/Projects/LongTermExpEvo/MethylationProject/Manuscript/Figures/figure_S2.png",
#       width = 6,
#       height = 4,
#       dpi = 320)
#
#ggsave(figure_s2,
#       filename = "/Users/carlstone/Library/CloudStorage/Box-Box/Behringer_Lab_Box_Drive/Projects/LongTermExpEvo/MethylationProject/Manuscript/Figures/figure_S2.pdf",
#       width = 6,
#       height = 4)
#

```


```{r anc_stats}
anc_kw <- WT_average_annotated_long %>% 
  kruskal_test(m ~ feature_type)
anc_dunn <- WT_average_annotated_long %>% 
  dunn_test(m ~ feature_type)
anc_wilcoxon_one_sample <- WT_average_annotated_long %>% 
  group_by(feature_type) %>% 
  wilcox_test(m ~ 1,
              mu = median(WT_average_annotated$m))
anc_wilcoxon_one_sample$p.adj <- p.adjust(anc_wilcoxon_one_sample$p,
                                          method = "BH")
anc_wilcoxon_feature_name <- WT_average_annotated_long %>% 
  filter(feature_type != "Genes" &
           feature_type != "Transcription Units" &
           feature_type != "mRNA Binding Sites" &
           feature_type != "No Feature") %>% 
    separate(feature_name, sep = ' ', c('feature_name', NA)) %>%
  group_by(feature_name) %>% 
  wilcox_test(m ~ 1,
              mu = median(WT_average_annotated$m))
anc_wilcoxon_feature_name$p.adj <- p.adjust(anc_wilcoxon_feature_name$p,
                                            method = "BH")
anc_wilcoxon_feature_name$stars <- stars.pval(anc_wilcoxon_feature_name$p.adj)
anc_wilcoxon_feature_name[anc_wilcoxon_feature_name$p.adj < 0.05,]

anc_medians <- WT_average_annotated_long %>% 
  group_by(feature_type) %>% 
  summarise(median_methyl = median(beta),
            median_m = median(m))
anc_medians_feature <- WT_average_annotated_long %>% 
  separate(feature_name, sep = ' ', c('feature_name', NA)) %>%
  group_by(feature_name) %>% 
  summarise(median_methyl = median(beta),
            median_m = median(m))
anc_medians_feature[anc_medians_feature$feature_name %in% (genome_sites %>% filter(Type == "IS-Elements"))$Site,]

  
```


```{r anc_differences_by_feature}
WT_average_annotated_long$feature_type <- 
  as.factor(WT_average_annotated_long$feature_type)
levels(WT_average_annotated_long$feature_type) <- list('Promoters' = 'Promoters',
                                                        '-10 Box' = 'BOX_10',
                                                        '-35 Box' = 'BOX_35',
                                                        'Sigma 24' = 'Sigma24',
                                                        'Sigma 28' = 'Sigma28',
                                                        'Sigma 32' = 'Sigma32',
                                                        'Sigma 38' = 'Sigma38',
                                                        'Sigma 70' = 'Sigma70',
                                                        'DNA Binding Sites' = 
                                                          'DNA-Binding-Sites',
                                                        'Transcription Units' = 
                                                          'Transcription-Units',
                                                        'Genes' = 'Genes',
                                                        'Rho-Dependent Terminators' = 
                                                          'Rho-Dependent-Terminators',
                                                        'Rho-Independent Terminators' = 
                                                          'Rho-Independent-Terminators',
                                                        'mRNA Binding Sites' = 
                                                          'mRNA-Binding-Sites',
                                                        'Origin of Replication' = 
                                                          'Origin-of-Replication',          
                                                        'Cryptic Prophages' = 
                                                          'Cryptic-Prophages',
                                                        'IS Elements' = 'IS-Elements',
                                                        'REP Elements' = 'REP-Elements',
                                                        'No Feature' = 'No_Feature'
                                                        )

# All features
fig_s7_feature_methylation_all <- WT_average_annotated_long %>% 
  filter(!grepl('Transcription|Sigma|mRNA', feature_type)) %>% 
  dplyr::mutate(feature_type = case_when(
  feature_type == "No Feature" ~ "Intergenic\nNo Other Feature",
  feature_type == "Rho-Dependent Terminators" ~ "Rho-Dependent\nTerminators",
  feature_type == "Rho-Independent Terminators" ~"Rho-Independent\nTerminators",
  feature_type == "Origin of Replication" ~ "Origin of\nReplication",
  feature_type == "DNA Binding Sites" ~ "DNA Binding\nSites",
  feature_type == "Cryptic Prophages" ~ "Cryptic\nProphges",
  .default = feature_type)) %>% 
  dplyr::mutate(feature_type = fct_relevel(feature_type,
                                           "Genes",
                                           "Promoters",
                                           "-10 Box",
                                           "-35 Box",
                                           "DNA Binding\nSites",
                                           "Cryptic\nProphges",
                                           "Rho-Dependent\nTerminators",
                                           "Rho-Independent\nTerminators",
                                           "Origin of\nReplication",
                                           "IS Elements",
                                           "REP Elements",
                                           "Intergenic\nNo Other Feature")) %>%
  ggplot(aes(x = feature_type, y = beta)) +
  geom_boxplot(notch = F,
               fill = 'gray85') +
  theme_classic() +
  labs(x = 'Feature type',
       y = 'Percent methylated reads')

#ggview(fig_s7_feature_methylation_all,
#       width = 13,
#       height = 4,
#       units = "in")
#
#ggsave(plot = fig_s7_feature_methylation_all,
#       filename = "figure_s7.png",
#       path = "/Users/carlstone/Library/CloudStorage/Box-Box/Behringer_Lab_Box_Drive/Projects/LongTermExpEvo/MethylationProject/Manuscript/Figures",
#       width = 13,
#       height = 4,
#       units = "in")
#ggsave(plot = fig_s7_feature_methylation_all,
#       filename = "figure_s7.pdf",
#       path = "/Users/carlstone/Library/CloudStorage/Box-Box/Behringer_Lab_Box_Drive/Projects/LongTermExpEvo/MethylationProject/Manuscript/Figures",
#       width = 13,
#       height = 4,
#       units = "in")

fig_s7 <- WT_average_annotated_long |> 
  dplyr::mutate(feature_type = case_when(
    feature_type == "No Feature" ~ "Intergenic\nNo Other Feature",
    feature_type == "Rho-Dependent Terminators" ~ "Rho-Dependent\nTerminators",
    feature_type == "Rho-Independent Terminators" ~ "Rho-Independent\nTerminators",
    .default = feature_type)) |> 
  filter(!grepl('Sigma', feature_type)) %>% 
  filter(feature_type != 'Sigma 32') %>% 
  ggplot(aes(x = feature_type,
             y = beta)) +
  geom_boxplot(fill = "gray85") +
  theme_classic()

# Just sigma factors
comparison_list <- rev(list(c('Sigma 24','Sigma 38'),
                            c('Sigma 24','Sigma 70'),
                            c('Sigma 28','Sigma 38'),
                            c('Sigma 28','Sigma 70')))
sigma_avg_methylation_plot <- WT_average_annotated_long %>% 
  filter(grepl('Sigma', feature_type)) %>% 
  filter(feature_type != 'Sigma 32') %>% 
  ggplot(aes(x = feature_type, y = beta*100)) +
  geom_boxplot(notch = FALSE,
               fill = 'gray85') +
  theme_classic() +
  labs(x = 'Sigma factor binding at -35 element',
       y = "Percent methylated reads") +
  geom_hline(yintercept = median(WT_average$beta*100),
             linetype = "dashed")
sigma_avg_methylation_plot

# Number of GATC sites in each Sigma factor binding site
WT_average_annotated_long %>% 
  filter(grepl('Sigma', feature_type)) %>% 
  group_by(feature_type) %>%
  summarise(n = n())


  

# Just IS elements
IS_avg_methylation_plot <- WT_average_annotated_long %>% 
  filter(feature_type == "IS Elements") %>% 
  ggplot(aes(x = feature_name, y = beta)) +
  geom_boxplot(notch = T,
               fill = 'gray85') +
  theme_classic() +
  labs(x = 'IS element',
       y = 'Percent methylated reads')
IS_avg_methylation_plot

# Just cryptic prophages
prophage_avg_methylation_plot <- WT_average_annotated_long %>% 
  filter(feature_type == 'Cryptic Prophages') %>% 
  ggplot(aes(x = feature_name, y = beta)) +
  geom_boxplot(notch = T,
               fill = 'gray85') +
  theme_classic() +
  labs(x = 'Cryptic prophage',
       y = 'Percent methylated reads')
prophage_avg_methylation_plot

# Just DNA binding sites
DBS_avg_methylation_plot <- WT_average_annotated_long %>%
  filter(feature_type == 'DNA Binding Sites') %>% 
  separate(feature_name, sep = ' ', c('feature_name', NA)) %>% 
  ggplot(aes(x = feature_name, y = beta)) +
  geom_boxplot(notch = T,
               fill = 'gray85') +
  theme_classic() +
  labs(x = 'DNA Binding Factor',
       y = 'Percent methylated reads')

# Summary table of various DNA binding sites
all_DBS <- WT_average_annotated_long %>%
  filter(feature_type == 'DNA Binding Sites') %>% 
  separate(feature_name, sep = ' ', c('feature_name', NA)) %>% 
  group_by(feature_name) %>% 
  count(feature_name)

# Just DNA binding sites, excluding those with <3 GATC sites
DBS_threeplus_avg_methylation_plot <- WT_average_annotated_long %>%
  filter(feature_type == 'DNA Binding Sites') %>% 
  separate(feature_name, sep = ' ', c('feature_name', NA)) %>% 
  filter(feature_name %in% all_DBS[all_DBS$n > 2,]$feature_name) %>% 
  ggplot(aes(x = feature_name, y = beta)) +
  geom_boxplot(notch = F,
               fill = 'gray85') +
  theme_classic() +
  labs(x = 'DNA Binding Factor',
       y = 'Percent methylated reads')
DBS_threeplus_avg_methylation_plot

# REP elements
REP_avg_methylation_plot <- WT_average_annotated_long %>% 
  filter(feature_type == 'REP Elements') %>% 
  ggplot(aes(x = feature_name, y = beta)) +
  geom_boxplot(notch = F,
               fill = 'gray85') +
  theme_classic() +
  labs(x = 'REP Element',
       y = 'Percent methylated reads')
REP_avg_methylation_plot

# Genic vs intergenic sites
genic_df <- WT_average_annotated %>% 
  filter(!is.na(Genes)) %>% 
  dplyr::select(Position, beta)
intergenic_df <- WT_average_annotated %>% 
  filter(is.na(Genes)) %>% 
  dplyr::select(Position, beta)
genic_df$gi <- "genic"
intergenic_df$gi <- "intergenic"
genic_intergenic_df <- bind_rows(genic_df, intergenic_df)
genic_plot <- ggplot(genic_intergenic_df,
                     aes(x = gi,
                         y = beta)) +
  geom_boxplot(notch = T, fill = 'gray85') +
  theme_classic() +
  labs(x = 'Genic/Intergenic',
       y = 'Percent methylated reads')
genic_plot

genic_intergenic_df$m <- methylBtoM(genic_intergenic_df$beta)
genic_intergenic_wilcox <- genic_intergenic_df %>% 
  wilcox_test(m ~ gi, exact = TRUE, detailed = TRUE)
```

```{r figure_3_beta_at_genomic_features}
# Panel A: TSS
WT_all_TSS <- annotateTSS(WT_average_annotated, 
                          genome_sites, 
                          location = 'Position', 
                          size = 500)
WT_all_TSS <- WT_all_TSS %>% pivot_longer(cols = starts_with('RelPos'),
                                          names_to = 'TSS_strand',
                                          names_pattern = 'RelPos_(.)[0-9]*',
                                          values_to = 'RelPos') %>% 
  distinct(Position, RelPos, .keep_all = TRUE) %>% 
  filter(!is.na(RelPos)) %>% 
  dplyr::select(c(Position:m,NoTSS:RelPos))

lambdavar <- 1000

TSS_plot <- ggplot(data = WT_all_TSS, aes(x = RelPos, y = beta*100)) +
  stat_summary(geom = 'point',
               fun = median) +
  geom_quantile(quantiles = 0.5,
                method = 'rqss',
                lambda = lambdavar,
                linewidth = 1) +
  geom_vline(xintercept = -35, linetype = 'dashed', color = 'red', linewidth = 1) +
  theme_classic() +
  labs(x = 'Position relative to TSS',
       y = "Percent methylated reads")
TSS_plot

TSS_cov <- ggplot(data = WT_all_TSS, aes(x = RelPos, y = cov)) +
  stat_summary(geom = "point",
               fun = median) +
  geom_quantile(quantiles = 0.5,
                method = 'rqss',
                lambda = lambdavar,
                linewidth = 1)

# Use regression m ~ cov to predict TSS sites based on cov
# Shows that effect of coverage on m here is negligible
fit <- lm(m ~ cov, data = WT_average)
WT_all_TSS <- bind_cols(WT_all_TSS, predict(fit, WT_all_TSS, interval = 'confidence'))
WT_all_TSS |> 
  group_by(RelPos) |> 
  summarise(cov = mean(cov),
            beta = mean(beta),
            m = mean(m)) |> 
  ggplot(aes(x = RelPos, y = m, color = cov)) +
  geom_point() +
  scale_color_continuous(type = "viridis")

# Panel B: Sigma factors
# Use sigma_avg_methylation_plot from chunk above

# Panel C: Methylation in other genomic features
# Big figure with only the hits for main figure
# Set whether you want notches on all of the plots here
notch_bool <- FALSE
genome_median <- median(WT_average$beta*100)
genic_combined_plot <- ggplot(genic_intergenic_df,
                     aes(x = gi,
                         y = beta*100)) +
  geom_boxplot(notch = notch_bool,
               fill = '#FBB4AE') +
  theme_classic() +
  labs(x = 'Genic/Intergenic',
       y = "Percent methylated reads") +
  scale_y_continuous(breaks = seq(0, 100, 20),
                     limits = c(0,100)) +
  geom_hline(yintercept = genome_median,
             linetype = 'dashed') +
  theme(plot.margin = unit(c(0.1, 0, 0.1, 0.1), "cm"))
DBS_combined_plot <- WT_average_annotated_long %>%
  filter(feature_type == 'DNA Binding Sites') %>% 
  separate(feature_name, sep = ' ', c('feature_name', NA)) %>% 
  filter(feature_name == 'CRP-cAMP' |
           feature_name == 'Fnr' |
           feature_name == 'Nac' |
           feature_name == 'Fis' |
           feature_name == 'Cra' |
           feature_name == 'IHF' |
           feature_name == 'Lrp') %>% 
  dplyr::mutate(feature_name = forcats::fct_recode(feature_name, 'CRP' = 'CRP-cAMP')) %>% 
#  dplyr::mutate(feature_type = 
#                  forcats::fct_relevel(feature_type,
#                                       ))
  ggplot(aes(x = feature_name, y = beta*100)) +
  geom_boxplot(notch = notch_bool,
               fill = '#B3CDE3') +
  theme_classic() +
  labs(x = 'DNA Binding Factors') +
  scale_y_continuous(limits = c(0,100)) +
  theme(axis.line.y = element_blank(), 
        axis.text.y = element_blank(), 
        axis.ticks.y = element_blank(), 
        axis.title.y = element_blank(),
        plot.margin = unit(c(0.1, 0, 0.1, 0), "cm")) +
  geom_hline(yintercept = genome_median,
             linetype = 'dashed')
prophage_combined_plot <- WT_average_annotated_long %>% 
  filter(feature_name == 'DLP12' |
           feature_name == 'Rac' |
           feature_name == 'Qin' |
           feature_name == 'e14' |
           feature_name == 'CPZ-55') %>%
  ggplot(aes(x = feature_name, y = beta*100)) +
  geom_boxplot(notch = notch_bool,
               fill = '#CCEBC5') +
  theme_classic() +
  labs(x = 'Cryptic Prophages') +
  scale_y_continuous(limits = c(0,100)) +
  theme(axis.line.y = element_blank(), 
        axis.text.y = element_blank(), 
        axis.ticks.y = element_blank(), 
        axis.title.y = element_blank(),
        plot.margin = unit(c(0.1, 0, 0.1, 0), "cm")) +
  geom_hline(yintercept = genome_median,
             linetype = 'dashed')
IS_combined_plot <- WT_average_annotated_long %>% 
  filter(feature_name == 'IS2A' |
           feature_name == 'IS2D' |
           feature_name == 'IS5A' |
           feature_name == 'IS5B' |
           feature_name == 'IS150' |
           feature_name == 'IS186A' |
           feature_name == 'IS186B') %>% 
  ggplot(aes(x = feature_name, y = beta*100)) +
  geom_boxplot(notch = notch_bool,
               fill = '#DECBE4') +
  theme_classic() +
  labs(x = 'IS Elements') +
  scale_y_continuous(limits = c(0,100)) +
  theme(axis.line.y = element_blank(), 
        axis.text.y = element_blank(), 
        axis.ticks.y = element_blank(), 
        axis.title.y = element_blank(),
        plot.margin = unit(c(0.1, 0, 0.1, 0), "cm")) +
  geom_hline(yintercept = genome_median,
             linetype = 'dashed')

# Put it together
combined_plot <- plot_grid(genic_combined_plot, DBS_combined_plot, prophage_combined_plot,
                           IS_combined_plot,
                           nrow = 1,
                           rel_widths = c(5/9, 1, 5/7, 1),
                           labels = c("C", NULL, NULL, NULL, NULL))

bottom_plot <- plot_grid(NULL, combined_plot,
                         nrow = 2,
                         rel_heights = c(0.05, 1))

fig3_top <- plot_grid(TSS_plot, sigma_avg_methylation_plot,
                      nrow = 1,
                      labels = c("A", "B"))

figure_3 <- plot_grid(fig3_top, bottom_plot,
                      nrow = 2,
                      align = "v",
                      axis = "r")
figure_3

ggview(figure_3, width = 10, height = 7)


#ggsave(plot = figure_3,
#       filename = "figure_3.png",
#       path = "/Users/carlstone/Library/CloudStorage/Box-Box/Behringer_Lab_Box_Drive/Projects/LongTermExpEvo/MethylationProject/Manuscript/Figures",
#       width = 10,
#       height = 7,
#       dpi = 320)
#ggsave(plot = figure_3,
#       filename = "figure_3.pdf",
#       path = "/Users/carlstone/Library/CloudStorage/Box-Box/Behringer_Lab_Box_Drive/Projects/LongTermExpEvo/MethylationProject/Manuscript/Figures",
#       width = 10,
#       height = 7)

#ggsave(plot = figure_3,
#       filename = "figure_3.svg",
#       path = "/Users/carlstone/Library/CloudStorage/Box-Box/Behringer_Lab_Box_Drive/Projects/LongTermExpEvo/MethylationProject/Manuscript/Figures",
#       width = 10,
#       height = 7)
```

## Comparative analysis

### Summary differences between samples

```{r comparative_summary_stats}
summary_by_genotype <- mgAllSamplesLong %>% 
  group_by(genotype) %>% 
  summarize(med_beta = median(beta),
            med_m = median(m))
summary_by_sample <-  mgAllSamplesLong %>% 
  group_by(sample) %>% 
  summarize(med_beta = median(beta),
            med_m = median(m))

evolvedWT_average <- mgAllSamplesLong %>% 
  filter(sample %in% c("125.6", "125.7","129.1", "129.3")) %>% 
  group_by(Position) %>% 
  summarise(beta = mean(beta),
            m = mean(m))
evolvedMMR_average <- mgAllSamplesLong %>% 
  filter(sample %in% c("113.1", "113.2","126.2", "126.4")) %>% 
  group_by(Position) %>% 
  summarise(beta = mean(beta),
            m = mean(m))

all_averages <- WT_average %>% 
  left_join(evolvedWT_average, by = "Position", suffix = c("_anc", "_WT")) %>% 
  left_join(evolvedMMR_average, by = "Position") %>% 
  dplyr::rename(beta_MMR = beta, m_MMR = m) %>% 
  drop_na()

all_averages$WT_delta <- all_averages$beta_WT - all_averages$beta_anc
all_averages$MMR_delta <- all_averages$beta_MMR - all_averages$beta_anc

wilcox.test(all_averages$m_WT, all_averages$m_MMR)
evolved_medians <- mgAllSamplesLong %>% 
  group_by(genotype) %>% 
  summarise(median = median(m))

plot_data <- tibble(genotype = c(rep("WT", length(all_averages$Position)),
                                 rep("MMR-", length(all_averages$Position))),
                    beta = c(all_averages$WT_delta,
                             all_averages$MMR_delta))

figure_s4 <- plot_data %>% 
  ggplot(aes(x = beta*100, color = genotype, fill = genotype)) +
  stat_ecdf() +
  coord_cartesian(xlim = c(-15, 15)) +
  scale_color_manual("Genotype", values = c("#377EB8", "#E6E61A")) +
  theme_stone() +
  geom_vline(xintercept = 0) +
  labs(x = "Change in percent methylation",
       y = "Cumulative density")

ggview(figure_s4,
       width = 6,
       height = 4)

#ggsave(plot = figure_s4,
#       filename = "/Users/carlstone/Library/CloudStorage/Box-Box/Behringer_Lab_Box_Drive/Projects/LongTermExpEvo/MethylationProject/Manuscript/Figures/figure_s4.pdf",
#       width = 6,
#       height = 4)
  
#ggsave(plot = figure_s4,
#       filename = "/Users/carlstone/Library/CloudStorage/Box-Box/Behringer_Lab_Box_Drive/Projects/LongTermExpEvo/MethylationProject/Manuscript/Figures/figure_s4.png",
#       width = 6,
#       height = 4,
#       dpi = 320)

# Do KW test and Dunn's test
sample_distribution_plots <- mgAllSamplesLong %>% 
  ggplot(aes(x = beta,
             color = sample)) +
  stat_ecdf()
  

evolved_kw <- mgAllSamplesLong %>% 
  kruskal_test(m ~ genotype)
evolved_dunn <- mgAllSamplesLong %>% 
  dunn_test(m ~ genotype)
evolved_dunn_samples <- mgAllSamplesLong %>% 
  dunn_test(m ~ sample)

anc_kw <- WT_average_annotated_long %>% 
  kruskal_test(m ~ feature_type)
anc_dunn <- WT_average_annotated_long %>% 
  dunn_test(m ~ feature_type)
anc_wilcoxon_one_sample <- WT_average_annotated_long %>% 
  group_by(feature_type) %>% 
  wilcox_test(m ~ 1,
              mu = median(WT_average_annotated$m))
anc_wilcoxon_one_sample$p.adj <- p.adjust(anc_wilcoxon_one_sample$p,
                                          method = "BH")
anc_wilcoxon_one_sample$stars <- stars.pval(anc_wilcoxon_one_sample$p.adj)
anc_wilcoxon_feature_name <- WT_average_annotated_long %>% 
  filter(feature_type != "Genes" &
           feature_type != "Transcription Units" &
           feature_type != "mRNA Binding Sites" &
           feature_type != "No Feature") %>% 
  group_by(feature_name) %>% 
  wilcox_test(m ~ 1,
              mu = median(WT_average_annotated$m))
anc_wilcoxon_feature_name$p.adj <- p.adjust(anc_wilcoxon_feature_name$p,
                                            method = "BH")
anc_wilcoxon_feature_name$stars <- stars.pval(anc_wilcoxon_feature_name$p.adj)
#View(anc_wilcoxon_feature_name[anc_wilcoxon_feature_name$p.adj < 0.05,])

```


### Differential Methylation

Need to use getData and methylRaw methods to build methylRaw objects from mgAllSamples_normalized

methylRaw object has slots
sample.id = "WT1"
assembly =  "NC_000913.3"
context = "GATC"
resolution = "base"
names = "chr" "start" "end" "strand" "coverage" "numCs" "numTs"
row.names which are just 1:end
.Data which is a list [[1]] through [[7]] of the columns in names

```{r methylkit_data_processing}

# TODO Join mgAllSamples_normalized to original df with strand +/- and input that
# into the function

dm_data <- mgAllSamples[,2:24]
# Quantile normalize coverage values between samples
#dm_data[,2:12] <- dm_data %>% 
#  dplyr::select(starts_with("cov")) %>% 
#  as.matrix() %>% 
#  preprocessCore::normalize.quantiles(x = .)

dmObject_raw <- buildMethylRawList(as.data.frame(dm_data))
#dmObject@treatment = c(0,0,0,2,2,1,1,2,2,1,1)
dmObject_raw@treatment = c(0, 0, 0, 2, 2, 1, 1, 2, 2, 1, 1)
dmObject <- methylKit::normalizeCoverage(dmObject_raw, method = 'median')

```

```{r run_differential_methylation}
allMethyl <- methylKit::unite(dmObject)


methylAvWT <- reorganize(allMethyl,
                           sample.ids = c('WT1', 'WT2', 'WT3', 
                                          '125.6', '125.7', '129.1', '129.3'),
                           treatment = c(0,0,0,1,1,1,1))
methylAvMMR <- reorganize(allMethyl,
                            sample.ids = c('WT1', 'WT2', 'WT3', 
                                           '113.1', '113.2', '126.2', '126.4'),
                            treatment = c(0,0,0,2,2,2,2))
suppressWarnings(
dmAvWT <- calculateDiffMeth(methylAvWT,
                            overdispersion = 'none',
                            effect = 'wmean')
)
suppressWarnings(
dmAvMMR <- calculateDiffMeth(methylAvMMR,
                             overdispersion = 'none',
                             effect = 'wmean')
)

# this method uses glm on the back end to do logistic regression, but
# unfortunately it gives a warning for every site because it is a 0-1 decimal
# and not an integer, so use suppressWarnings if you don't want 38000 warnings
#suppressWarnings(
#dmAll <- calculateDiffMeth(allMethyl,
#                           overdispersion = "none",
#                           effect = "wmean")
#)

# dm_tibble <- as_tibble(dmAll)

dmAvWT <- as_tibble(dmAvWT)
dmAvMMR <- as_tibble(dmAvMMR)

# Save for later so we don't have to do that again
#write_tsv(dm_tibble,
#          file = "/Users/carlstone/Library/CloudStorage/Box-Box/Behringer_Lab_Box_Drive/Projects/LongTermExpEvo/MethylationProject/Rscripts/diffMethSites.txt")

#write_tsv(dmAvWT,
#          file = "/Users/carlstone/Library/CloudStorage/Box-Box/Behringer_Lab_Box_Drive/Projects/LongTermExpEvo/MethylationProject/Rscripts/dmAvWT.txt")
#write_tsv(dmAvMMR,
#          file = "/Users/carlstone/Library/CloudStorage/Box-Box/Behringer_Lab_Box_Drive/Projects/LongTermExpEvo/MethylationProject/Rscripts/dmAvMMR.txt")
```


```{r dm_analysis_all, eval=FALSE, include=FALSE}
dm_tibble <- read_tsv(file = "/Users/carlstone/Library/CloudStorage/Box-Box/Behringer_Lab_Box_Drive/Projects/LongTermExpEvo/MethylationProject/Rscripts/diffMethSites.txt")

dmAvWT<- read_tsv(file = "/Users/carlstone/Library/CloudStorage/Box-Box/Behringer_Lab_Box_Drive/Projects/LongTermExpEvo/MethylationProject/Rscripts/dmAvWT.txt")

dmAvMMR<- read_tsv(file = "/Users/carlstone/Library/CloudStorage/Box-Box/Behringer_Lab_Box_Drive/Projects/LongTermExpEvo/MethylationProject/Rscripts/dmAvMMR.txt")

siteMutations <- read_tsv(file = '/Users/carlstone/Library/CloudStorage/Box-Box/Behringer_Lab_Box_Drive/Projects/LongTermExpEvo/MethylationProject/GATC_Mutation_Data/SummaryData/GATCsiteSummary.txt')

dm_metadata <- annotateMethylSites(dm_tibble,
                                   genome_sites,
                                   location = "start")

dm_metadata$mutated <- "NO"
mutatedGATCs <- rep(siteMutations$position, each = 7) + seq(-3,3)
dm_metadata <- dm_metadata %>% 
  filter(!(start %in% mutatedGATCs))

percentCutoff <- 10
qvalueCutoff <- 0.05

# Classify as differentially methylated up or down for later coloring
dm_metadata$diffexpressed <- "NO"
dm_metadata$diffexpressed[dm_metadata$meth.diff > percentCutoff &
                            dm_metadata$qvalue < qvalueCutoff] <- "UP"

mycolors <- c("red","black")
names(mycolors) <- c("UP","NO")

dm_metadata %>% 
  ggplot(aes(x = meth.diff,
             y = -log10(qvalue),
             col = diffexpressed)) +
  geom_point() +
  scale_color_manual(values = mycolors) +
  geom_hline(yintercept = -log10(qvalueCutoff),
             col = 'red',
             linetype = 'dashed') +
  theme_stone()

```


```{r dm_analysis_split_by_group}

dmAvWT <- read_tsv(file = "/Users/carlstone/Library/CloudStorage/Box-Box/Behringer_Lab_Box_Drive/Projects/LongTermExpEvo/MethylationProject/Rscripts/dmAvWT.txt")

dmAvMMR <- read_tsv(file = "/Users/carlstone/Library/CloudStorage/Box-Box/Behringer_Lab_Box_Drive/Projects/LongTermExpEvo/MethylationProject/Rscripts/dmAvMMR.txt")

dmWT_metadata <- annotateMethylSites(dmAvWT,
                                     genome_sites,
                                     location = "start")
dmMMR_metadata <- annotateMethylSites(dmAvMMR,
                                     genome_sites,
                                     location = "start")

dmWT_metadata$mutated <- "NO"
dmMMR_metadata$mutated <- "NO"
mutatedGATCs <- rep(siteMutations$position, each = 7) + seq(-3,3)
dmWT_metadata <- dmWT_metadata %>% 
  filter(!(start %in% mutatedGATCs))
dmMMR_metadata <- dmMMR_metadata %>% 
  filter(!(start %in% mutatedGATCs))

percentCutoff <- 10
qvalueCutoff <- 0.05

# Classify as differentially methylated up or down for later coloring
dmWT_metadata$diffexpressed <- "No"
dmWT_metadata$diffexpressed[dmWT_metadata$meth.diff > percentCutoff &
                            dmWT_metadata$qvalue < qvalueCutoff] <- "Up"
dmWT_metadata$diffexpressed[dmWT_metadata$meth.diff < -percentCutoff &
                            dmWT_metadata$qvalue < qvalueCutoff] <- "Down"
dmMMR_metadata$diffexpressed <- "No"
dmMMR_metadata$diffexpressed[dmMMR_metadata$meth.diff > percentCutoff &
                             dmMMR_metadata$qvalue < qvalueCutoff] <- "Up"
dmMMR_metadata$diffexpressed[dmMMR_metadata$meth.diff < -percentCutoff &
                             dmMMR_metadata$qvalue < qvalueCutoff] <- "Down"

mycolors <- c("red", "blue", "gray")
names(mycolors) <- c("Up", "Down","No")

dmWT_plot <- dmWT_metadata %>% 
  ggplot(aes(x = meth.diff,
             y = -log10(qvalue),
             col = diffexpressed)) +
  geom_point() +
  scale_color_manual(values = mycolors) +
  geom_hline(yintercept = -log10(qvalueCutoff),
             col = "black",
             linetype = "dashed") +
  geom_vline(xintercept = c(-10, 10),
             col = "black",
             linetype = "dashed") +
  labs(x = "% methylation difference") +
  lims(x = c(-25, 25),
       y = c(NA, 25)) +
  theme_stone() +
  theme(panel.background = element_rect(color = "black",
                                        linewidth = 1),
        axis.line.x.bottom = element_line(linewidth = 0),
        axis.line.y.left = element_line(linewidth = 0)) +
  coord_fixed(ratio = 2) +
  annotate(
    'text',
    x = c(-19, 19),
    y = c(19, 19),
    label = c('16', '22'),
    size = c(7, 7),
    color = c('blue', 'red')
  )

dmMMR_plot <- dmMMR_metadata %>% 
  ggplot(aes(x = meth.diff,
             y = -log10(qvalue),
             col = diffexpressed)) +
  geom_point() +
  scale_color_manual(values = mycolors) +
  geom_hline(yintercept = -log10(qvalueCutoff),
             col = "black",
             linetype = "dashed") +
  geom_vline(xintercept = c(-10, 10),
             col = "black",
             linetype = "dashed") +
  labs(x = "% methylation difference") +
  lims(x = c(-25, 25),
       y = c(NA, 25)) +
  theme_stone() +
#  theme(panel.background = element_rect(color = "black",
#                                        linewidth = 1),
#        axis.line.x.bottom = element_line(linewidth = 0),
#        axis.line.y.left = element_line(linewidth = 0),
#        legend.position = "none") +
    theme(legend.position = "none") +
  coord_fixed(ratio = 2) +
  annotate(
    'text',
    x = c(-19, 19),
    y = c(19, 19),
    label = c('141', '28'),
    size = c(7, 7),
    color = c('blue', 'red')
  )

dm_legend <- cowplot::get_legend(dmWT_plot + theme(legend.box.margin = margin(0, 0,
                                                                              0, 12)))

dm_combined <- plot_grid(NULL, dmWT_plot + theme(legend.position = "none"), 
                         NULL, dmMMR_plot,
                         ncol = 1,
                         rel_heights = c(0.15, 1, -0.01, 1),
                         label_x = 0,
                         label_y = 1,
                         labels = c("","A","","B"))
dm_combined
#ggview(dm_combined, width = 2.5, height = 4.5)
```

### Ordination

```{r normalize_data_before_ordination}
#mgAllSamples_normalized_list <- methylationNormalizeQuantiles(df = mgAllSamples,
#                                                              normalize_position = TRUE,
#                                                              plots = TRUE,
#                                                              shift_pos = FALSE,
#                                                              rescale = TRUE)
mgAllSamples_normalized_list <- methylationNormalize(df = mgAllSamples,
                                                     normalize_position = TRUE,
                                                     plots = FALSE,
                                                     rescale = TRUE)
mgAllSamples_normalized <- mgAllSamples_normalized_list$data

# Check how well samples were normalized
# plot_grid(plotlist = mgAllSamples_normalized_list$cov_plots)
# plot_grid(plotlist = mgAllSamples_normalized_list$methyl_plots)


mgAllSamples_normalized_long <- mgAllSamples_normalized %>% 
  pivot_longer(cols = !Position,
               names_to = c('.value', 'sample'),
               names_sep = '_')

mgAllSamples_normalized_long <- mgAllSamples_normalized_long %>% 
  dplyr::rename(beta = 'methyl')

mgAllSamples_normalized_long$m <- methylBtoM(mgAllSamples_normalized_long$beta,
                                             alpha = 1 / (max(mgAllSamples_normalized_long$cov) + 1))

mgAllSamples_normalized_long$genotype <- mgAllSamples_normalized_long$sample
mgAllSamples_normalized_long <- mgAllSamples_normalized_long %>% 
  mutate(genotype = recode(
    genotype,
    WT1 = 'anc',
    WT2 = 'anc',
    WT3 = 'anc',
    '125.6' = 'WT',
    '125.7' = 'WT',
    '129.1' = 'WT',
    '129.3' = 'WT',
    '113.1' = 'MMR',
    '113.2' = 'MMR',
    '126.2' = 'MMR',
    '126.4' = 'MMR'
  ))
```


```{r ordination}
# PCA with beta values and m values ---------------------------------------

pca_data <- mgAllSamples_normalized_long %>% 
  dplyr::select(!genotype) %>% 
  pivot_wider(names_from = sample,
              values_from = c(cov, beta, m),
              values_fill = NA) %>% 
  drop_na()

  

pca_m <- pca_data %>% 
  dplyr::select(starts_with('m')) %>% 
  t() %>% 
  prcomp(x = .)

pca_beta <- pca_data %>% 
  dplyr::select(starts_with('beta')) %>% 
  t() %>% 
  prcomp(x = .)

#autoplot(pca_beta, label = T)
#autoplot(pca_m, label = T)

# PCA too many dimensions so doing NMDS to flatten it


# NMDS
sites <- pca_data$Position
beta_distance <- pca_data %>% 
  dplyr::select(starts_with('beta')) %>% 
  t() %>% 
  vegdist(x = .,
          method = 'euclidean')
m_distance <- pca_data %>% 
  dplyr::select(starts_with('m')) %>% 
  t() %>% 
  vegdist(x = .,
          method = 'euclidean')
set.seed(1)
NMDS <- metaMDS(beta_distance,
                distance = 'euclidean',
                k = 3,
                trymax = 1000,
                maxit = 9999,
                noshare = F,
                zerodist = 'ignore',
                pc = T,
                center = T)
stressplot(NMDS)
# stress_values <- numeric(4)
# for (i in 1:4) {
#   NMDS <- metaMDS(beta_distance,
#                 distance = 'euclidean',
#                 model = 'linear',
#                 k = i,
#                 try = 250,
#                 maxit = 999,
#                 wascores = T,
#                 autotransform = F,
#                 expand = F,
#                 noshare = F,
#                 zerodist = 'ignore',
#                 pc = T,
#                 center = T)
#   stress_values[i] <- NMDS$stress
# }
# plot(x = 1:4, y = stress_values)
groups_NMDS <- rep(NA, 11)
groups_NMDS[1:3] <- 'Ancestor'
groups_NMDS[c(4,5,8,9)] <- 'MMR-'
groups_NMDS[c(6,7,10,11)] <- 'WT'
samples_NMDS <- c("Anc1","Anc2","Anc3","113.1","113.2","125.6","125.7","126.2","126.4","129.1","129.3")
#ggplot it
NMDS_scores <- as.data.frame(NMDS[['points']])
NMDS_scores$Sample <- rownames(NMDS_scores)
NMDS_scores$Genotype <- groups_NMDS
#Point size 2 for NMDS.png
plot_NMDS <- ggplot(NMDS_scores,
                    aes(x = MDS1,
                        y = MDS2,
                        fill = Genotype,
                        shape = Genotype,
                        label = samples_NMDS)) +
  geom_point(size = 4) +
  #geom_text_repel() +
  theme_stone() +
  theme(legend.text = element_text(size = 12),
        legend.position = c(0.25, 0.75),
        legend.background = element_rect(fill = 'white', color = 'black'),
        legend.title = element_blank()) +
  labs(color = 'Group',
       x = "NMDS1",
       y = "NMDS2") +
  scale_fill_manual(values = c('black', "#377EB8", "#FFFF33")) +
  scale_shape_manual(values = c(21, 22, 24))

```

```{r hierarchical clustering}
#beta_tree <- hclust(beta_distance)
#m_tree <- hclust(m_distance)
#plot(beta_tree)
#plot(m_tree)
#
#library(ape)
##write.tree(as.phylo(beta_tree),
#           file = "/Users/carlstone/Library/CloudStorage/Box-Box/Behringer_Lab_Box_Drive/Projects/LongTermExpEvo/MethylationProject/Trees/methyl_tree")
```

```{r perMANOVA}
# permutational MANOVA using adonis2 from vegan
pca_metadata <- tibble(Genotype = groups_NMDS,
                       Population = c("Anc","Anc","Anc",
                                      "P113","P113","P125","P125",
                                      "P126","P126","P129","P129"))
permanova <- adonis2(beta_distance ~ Genotype,
                     data = pca_metadata,
                     method = "euclidean")
print(permanova)
pairwise.adonis2(beta_distance ~ Genotype,
                 data = pca_metadata,
                 method = "euclidean")
beta_disper <- betadisper(beta_distance, group = groups_NMDS)
anova(beta_disper)
TukeyHSD(beta_disper)
permutest(beta_disper, pairwise = TRUE)

```


```{r figure_4_DM_and_NMDS}
# Combine dm_combined from above with plot_NMDS
cow_NMDS <- plot_grid(NULL, plot_NMDS,
                      ncol = 1,
                      rel_heights = c(0.075, 1),
                      labels = c("", "C"))
figure_4 <- plot_grid(dm_combined, NULL, cow_NMDS,
                      rel_widths = c(2, 0.1, 4),
                      nrow = 1,
                      align = "h",
                      axis = "bt")
ggview(figure_4, width = 6.5, height = 4.5)
#ggsave(filename = "/Users/carlstone/Library/CloudStorage/Box-Box/Behringer_Lab_Box_Drive/Projects/LongTermExpEvo/MethylationProject/Manuscript/Figures/figure_4.png",
#       plot = figure_4,
#       width = 6.5,
#       height = 4.5,
#       dpi = 320)
#ggsave(filename = "/Users/carlstone/Library/CloudStorage/Box-Box/Behringer_Lab_Box_Drive/Projects/LongTermExpEvo/MethylationProject/Manuscript/Figures/figure_4.pdf",
#       plot = figure_4,
#       width = 6.5,
#       height = 4.5)
```


```{r upsetplot_code}
list2df <- function(inputList) {
    # ldf <- lapply(1:length(inputList), function(i) {
    ldf <- lapply(seq_len(length(inputList)), function(i) {
        data.frame(categoryID=rep(names(inputList[i]),
                                  length(inputList[[i]])),
                   Gene=inputList[[i]])
    })

    do.call('rbind', ldf)
}

plot_upset <- function (x, n = 10, color = "grey", ...) {
    df <- as.data.frame(x)
    id <- df$ID[1:n]
    des <- df$Description[1:n]
    glist <- geneInCategory(x)[id]
    names(glist) <- des
    d <- list2df(glist)
    res <- tibble::tibble(Description = split(d[, 1], d[, 2]))
    ggplot(res, aes_(x = ~Description)) +
      geom_bar(color = color, fill = color) + 
      theme_classic() +
      #theme_dose(font.size = 12) + 
      xlab(NULL) + 
      ylab(NULL) + 
      ggupset::scale_x_upset(order_by = "freq") +
      theme(text = element_text(size = 12))
}
```

# GO analysis

```{r GO_analysis_data_setup}
all_dm_sites <- read_tsv("/Users/carlstone/Library/CloudStorage/Box-Box/Behringer_Lab_Box_Drive/Projects/LongTermExpEvo/MethylationProject/ComparativeAnalysis/DifferentialMethylation/all_dm_sites.txt")
WT_dm_sites <- read_tsv("/Users/carlstone/Library/CloudStorage/Box-Box/Behringer_Lab_Box_Drive/Projects/LongTermExpEvo/MethylationProject/ComparativeAnalysis/DifferentialMethylation/diffMethAncVsWT_metadata.txt")
MMR_dm_sites <- read_tsv("/Users/carlstone/Library/CloudStorage/Box-Box/Behringer_Lab_Box_Drive/Projects/LongTermExpEvo/MethylationProject/ComparativeAnalysis/DifferentialMethylation/diffMethAncVsMMR_metadata.txt")

#Process all_dm_sites for GO analysis
all_dm_sites <- all_dm_sites |> 
  tidyr::separate_wider_delim(cols = genic_location,
                              delim = "-",
                              names = c("gene1", "gene2"),
                              too_few = "align_start")
gene1 <- all_dm_sites$gene1
gene2 <- all_dm_sites$gene2
all_dm_sites <- all_dm_sites |> 
  bind_rows(all_dm_sites) |> 
  dplyr::select(!c(gene1, gene2))
all_dm_sites$gene <- c(gene1, gene2)
all_dm_sites <- all_dm_sites |> 
  drop_na()

write_tsv(all_dm_sites,
          file = "/Users/carlstone/Library/CloudStorage/Box-Box/Behringer_Lab_Box_Drive/Projects/LongTermExpEvo/MethylationProject/ComparativeAnalysis/DifferentialMethylation/all_dm_sites_clean.txt")

#Process WT and MMR separately for GSEA
WT_dm_sites$start <- WT_dm_sites$start + 1
MMR_dm_sites$start <- MMR_dm_sites$start + 1
WT_dm_sites <- WT_dm_sites |> 
  dplyr::filter(mutated == "NO") |> 
  dplyr::select(start, strand, qvalue, meth.diff, location_types, genic_location) |> 
  tidyr::separate_wider_delim(cols = genic_location,
                              delim = "-",
                              names = c("gene1", "gene2"),
                              too_few = "align_start")
WT_gene1 <- WT_dm_sites$gene1
WT_gene2 <- WT_dm_sites$gene2
WT_dm_sites <- WT_dm_sites |> 
  bind_rows(WT_dm_sites) |> 
  dplyr::select(!c(gene1, gene2))
WT_dm_sites$gene <- c(WT_gene1, WT_gene2)
WT_dm_sites <- WT_dm_sites |> 
  drop_na() |> 
  dplyr::mutate(gene = case_when(
    gene == "sfmZ" ~ "fimZ",
    gene == "flc" ~ "crcB",
    gene == "ybfK" ~ "speF",
    gene == "rhsO" ~ "ybfO",
    gene == "rspR" ~ "ydfH",
    gene == "insA" ~ "insA5",
    gene == "rhmA" ~ "yfaU",
    gene == "rhmT" ~ "yfaV",
    gene == "rhmR" ~ "yfaX",
    gene == "trmN" ~ "yfiC",
    .default = gene
  )) |> 
  rowwise() |> 
  dplyr::mutate(m = methylBtoM(meth.diff/100))

MMR_dm_sites <- MMR_dm_sites |> 
  dplyr::filter(mutated == "NO") |> 
  dplyr::select(start, strand, qvalue, meth.diff, location_types, genic_location) |> 
  tidyr::separate_wider_delim(cols = genic_location,
                              delim = "-",
                              names = c("gene1", "gene2"),
                              too_few = "align_start")
MMR_gene1 <- MMR_dm_sites$gene1
MMR_gene2 <- MMR_dm_sites$gene2
MMR_dm_sites <- MMR_dm_sites |> 
  bind_rows(MMR_dm_sites) |> 
  dplyr::select(!c(gene1, gene2))
MMR_dm_sites$gene <- c(MMR_gene1, MMR_gene2)
MMR_dm_sites <- MMR_dm_sites |> 
  drop_na() |> 
  dplyr::mutate(gene = case_when(
    gene == "sfmZ" ~ "fimZ",
    gene == "flc" ~ "crcB",
    gene == "ybfK" ~ "speF",
    gene == "rhsO" ~ "ybfO",
    gene == "rspR" ~ "ydfH",
    gene == "insA" ~ "insA5",
    gene == "rhmA" ~ "yfaU",
    gene == "rhmT" ~ "yfaV",
    gene == "rhmR" ~ "yfaX",
    gene == "trmN" ~ "yfiC",
    .default = gene
  )) |> 
  rowwise() |> 
  dplyr::mutate(m = methylBtoM(meth.diff/100))
```

```{r GOanalysis}
# keytypes(org.EcK12.eg.db)

# Prepare gene ID's
## Start with all_dm_sites
entrez_genes <- bitr(all_dm_sites$gene, fromType = "ALIAS",
                     toType = "ENTREZID", OrgDb = "org.EcK12.eg.db")
alias_genes <- bitr(all_dm_sites$gene, fromType = "ALIAS",
                     toType = "SYMBOL", OrgDb = "org.EcK12.eg.db")
all_dm_sites <- all_dm_sites |> 
  left_join(entrez_genes, by = c("gene" = "ALIAS")) |> 
  left_join(alias_genes, by = c("gene" = "ALIAS"))

up_geneNames <- all_dm_sites |> 
  dplyr::filter(meth.diff > 0) |> 
  dplyr::select(gene, SYMBOL) |> 
  unlist()
dn_geneNames <- all_dm_sites |> 
  dplyr::filter(meth.diff < 0) |> 
  dplyr::select(gene, SYMBOL) |> 
  unlist()

geneList <- all_dm_sites$meth.diff
names(geneList) <- as.character(all_dm_sites$ENTREZID)
geneList <- sort(geneList, decreasing = TRUE)

genes <- names(geneList)
#genes <- unique(genes)

geneList_up <- geneList[geneList > 0]
geneList_dn <- geneList[geneList < 0]

## Now do WT and MMR all sites separately for GSEA
entrez_WT <- bitr(WT_dm_sites$gene, fromType = "ALIAS",
                  toType = "ENTREZID", OrgDb = "org.EcK12.eg.db", drop = FALSE)
alias_WT <- bitr(WT_dm_sites$gene, fromType = "ALIAS",
                  toType = "SYMBOL", OrgDb = "org.EcK12.eg.db", drop = FALSE)
WT_dm_sites <- WT_dm_sites |> 
  left_join(entrez_WT, by = c("gene" = "ALIAS")) |> 
  left_join(alias_WT, by = c("gene" = "ALIAS"))
WT_geneList_all <- WT_dm_sites$meth.diff
names(WT_geneList_all) <- as.character(WT_dm_sites$ENTREZID)
WT_geneList_all <- sort(WT_geneList_all, decreasing = TRUE)
WT_geneList <- WT_dm_sites |> 
  dplyr::filter(qvalue < 0.05,
                (meth.diff <= -10 | meth.diff >= 10)) |> 
  pull(meth.diff)
WT_geneNames <- WT_dm_sites |> 
  dplyr::filter(qvalue < 0.05,
                (meth.diff <= -10 | meth.diff >= 10)) |> 
  dplyr::select(gene, SYMBOL) |> 
  unlist()
names(WT_geneList) <- WT_dm_sites |> 
  dplyr::filter(qvalue < 0.05,
                (meth.diff <= -10 | meth.diff >= 10)) |> 
  pull(ENTREZID)
WT_geneList <- sort(WT_geneList, decreasing = TRUE)

entrez_MMR <- bitr(MMR_dm_sites$gene, fromType = "ALIAS",
                  toType = "ENTREZID", OrgDb = "org.EcK12.eg.db", drop = FALSE)
alias_MMR <- bitr(MMR_dm_sites$gene, fromType = "ALIAS",
                  toType = "SYMBOL", OrgDb = "org.EcK12.eg.db", drop = FALSE)
MMR_dm_sites <- MMR_dm_sites |> 
  left_join(entrez_MMR, by = c("gene" = "ALIAS")) |> 
  left_join(alias_MMR, by = c("gene" = "ALIAS"))
MMR_geneList_all <- MMR_dm_sites$m
names(MMR_geneList_all) <- as.character(MMR_dm_sites$ENTREZID)
MMR_geneList_all <- sort(MMR_geneList_all, decreasing = TRUE)
MMR_geneList <- MMR_dm_sites |> 
  dplyr::filter(qvalue < 0.05,
                (meth.diff <= -10 | meth.diff >= 10)) |> 
  pull(meth.diff)
MMR_geneNames <- MMR_dm_sites |> 
  dplyr::filter(qvalue < 0.05,
                (meth.diff <= -10 | meth.diff >= 10)) |> 
  dplyr::select(gene, SYMBOL) |> 
  unlist()
names(MMR_geneList) <- MMR_dm_sites |> 
  dplyr::filter(qvalue < 0.05,
                (meth.diff <= -10 | meth.diff >= 10)) |> 
  pull(ENTREZID)
MMR_geneList <- sort(MMR_geneList, decreasing = TRUE)


# Classify GO
ggo <- groupGO(gene = genes,
               OrgDb = org.EcK12.eg.db,
               ont = "BP",
               level = 3,
               readable = TRUE)

# GO
ego <- enrichGO(gene = genes,
                OrgDb = org.EcK12.eg.db,
                ont = "BP",
                readable = TRUE)
#WT_go <- enrichGO(gene = names(WT_geneList),
#                  OrgDb = org.EcK12.eg.db,
#                  ont = "BP",
#                  readable = TRUE,
#                  pvalueCutoff = 0.05,
#                  qvalueCutoff = 0.2)
#MMR_go <- enrichGO(gene = names(MMR_geneList),
#                   OrgDb = org.EcK12.eg.db,
#                   ont = "BP",
#                   readable = TRUE,
#                   pvalueCutoff = 0.05,
#                   qvalueCutoff = 0.2)

# GO UP/DN
go_up <- enrichGO(gene = names(geneList_up),
                  OrgDb = org.EcK12.eg.db,
                  ont = "BP",
                  readable = TRUE,
                  pvalueCutoff = 0.05,
                  qvalueCutoff = 0.05)
go_dn <- enrichGO(gene = names(geneList_dn),
                  OrgDb = org.EcK12.eg.db,
                  ont = "BP",
                  readable = TRUE,
                  pvalueCutoff = 0.05,
                  qvalueCutoff = 0.05)

# SIMPLIFY SIMPLFY SIMPLIFY
ego <- clusterProfiler::simplify(ego, by = "pvalue")
go_up <- clusterProfiler::simplify(go_up, by = "pvalue")
go_dn <- clusterProfiler::simplify(go_dn, cutoff = 0.5, by = "pvalue")

# Combine lists
ego@result$group <- "All"
go_up@result$group <- "More Methylated"
go_dn@result$group <- "Less Methylated"
go_everything <- bind_rows(as.data.frame(ego),
                           as.data.frame(go_up),
                           as.data.frame(go_dn))

# Plot upset plots
n <- nrow(go_everything)
id <- go_everything$ID
des <- go_everything$Description
glist <- c(geneInCategory(ego)[ego$ID],
           geneInCategory(go_up)[go_up$ID],
           geneInCategory(go_dn)[go_dn$ID])
names(glist) <- des
d <- list2df(glist)
res <- tibble::tibble(Description = split(d[, 1], d[, 2]))
res <- tibble(Gene = names(res$Description), Ont = res$Description)
res$evogroup <- NA
res <- res |> dplyr::mutate(evogroup = case_when(
    (Gene %in% WT_geneNames & Gene %in% MMR_geneNames) ~ "Both",
    (Gene %in% WT_geneNames & !(Gene %in% MMR_geneNames)) ~ "WT",
    (Gene %in% MMR_geneNames & !(Gene %in% WT_geneNames)) ~ "MMR-")) |> 
  dplyr::mutate(upordown = case_when(
    Gene %in% up_geneNames ~ "Increased",
    Gene %in% dn_geneNames ~ "Decreased"))
res$evogroup <- factor(res$evogroup, levels = c("WT", "MMR-", "Both"))
res$upordown <- factor(res$upordown, levels = c("Increased", "Decreased"))
# Only one gene was in both up and down so I'm just copying it and manually
# changing so it is both up and down
res <- res |>  dplyr::filter(Gene == "waaJ") |> bind_rows(res)
res[1,4] <- "Decreased"
# Remove redundant ontologies in each gene
res$Ont <- map(res$Ont, \(x) unique(x))


upset_plot <- ggplot(res, aes(x = Ont,
                fill = evogroup)) +
  facet_grid(rows = vars(upordown)) +
  geom_bar(color = "black") +
  scale_y_continuous(expand = expansion(mult = c(0, 0.05))) +
  scale_x_upset(sets = c("cell adhesion",
                         "cell adhesion involved in biofilm formation",
                         "cell aggregation",
                         "biofilm formation",
                         "aggregation of unicellular organisms",
                         "negative regulation of metabolic process",
                         "mRNA catabolic process",
                         "response to acidic pH",
                         "response to nutrient levels",
                         "response to virus",
                         "lipopolysaccharide core region biosynthetic process",
                         "lipopolysaccharide core region metabolic process",
                         "liposaccharide metabolic process",
                         "polysaccharide metabolic process",
                         "oligosaccharide biosynthetic process",
                         "oligosaccharide metabolic process")) +
  scale_fill_manual(values = c("#FFFF33", "#377EB8", "#4DAF4A")) +
  theme_classic() +
  theme(text = element_text(size = 12),
        plot.margin = unit(c(0.5,0.5,0.5,7.1), "cm"),
        panel.border = element_rect(color = "black",
                                    linewidth = 1.5,
                                    fill = NA),
        axis.title.x = element_blank(),
        axis.text.y = element_text(color = "black"),
        legend.position = c(0.66,0.85),
        legend.direction = "horizontal",
        legend.title = element_text(size = 10,
                                    face = "bold"),
        legend.background = element_rect(color = "black")) +
  theme_combmatrix(combmatrix.label.make_space = FALSE,
                   combmatrix.label.text = element_text(color = "black",
                                                        face = "bold")) +
  labs(fill = "Evolved\nGenotype")

ggview(upset_plot,
       width = 8,
       height = 5.5,
       units = "in")

#ggsave(plot = upset_plot,
#       filename = "/Users/carlstone/Library/CloudStorage/Box-Box/Behringer_Lab_Box_Drive/Projects/LongTermExpEvo/MethylationProject/Manuscript/Figures/upsetplot_GO.pdf",
#       height = 5.5,
#       width = 8,
#       units = "in")
#ggsave(plot = upset_plot,
#       filename = "/Users/carlstone/Library/CloudStorage/Box-Box/Behringer_Lab_Box_Drive/Projects/LongTermExpEvo/MethylationProject/Manuscript/Figures/upsetplot_GO.png",
#       height = 5.5,
#       width = 8,
#       units = "in")

# Which genes are involved in certain processes?
#res[grepl("mRNA", res$Ont),]
#res[grepl("pH", res$Ont),]
res$Ont <- unname(res$Ont)
res |>
  group_by(Ont) |> 
  summarise(genes = toString(Gene)) |> 
  rowwise() |> 
  mutate(Ont = toString(Ont),
         count = length(str_split_1(genes, ","))) |> 
  arrange(desc(count)) |> 
  write_tsv(file = "/Users/carlstone/Library/CloudStorage/Box-Box/Behringer_Lab_Box_Drive/Projects/LongTermExpEvo/MethylationProject/Manuscript/GO_enriched_genes_table.txt")


upset_up <- upsetplot(go_up,  color = "blue")
upset_dn <- upsetplot(go_dn)
upset_all <- upsetplot(ego)
upset_updown <- plot_grid(upset_up, upset_dn,
                          ncol = 1)

#upsetplot(ego)
#
#ridgeplot(ego2, core_enrichment = FALSE)
#ridgeplot(ego)
```

# DM intersect euler plots

```{r dm_intersect_euler}
# Require library eulerr

# 38 dm sites in WT --> group A
# 169 dm sites in MMR- --> group B
# 13 within the same region --> A&B
# 10 of those within same GATC --> A&B&C
# 7 of those in same adenine --> A&B&C&D

dm_overlaps <- euler(c("A" = 25,
                       "B" = 156,
                       "A&B" = 13,
                       "A&B&C" = 10,
                       "A&B&C&D" = 7),
                     shape = "ellipse")
error_plot(dm_overlaps)

dm_euler <- plot(dm_overlaps,
                 quantities = TRUE,
                 labels = c("WT", "MMR-", "Same GATC", "Same Adenine"),
                 adjust_labels = TRUE)

# Make a bunch and pick the one that looks good
nplots <- 30
plot_list <- vector(mode = "list", length = nplots)
for (i in 1:nplots) {
  eulerobj <- euler(c("A" = 25,
                      "B" = 156,
                      "A&B" = 13,
                      "A&B&C" = 10,
                      "A&B&C&D" = 7),
                      shape = "ellipse")
  plot_list[[i]] <- as_ggplot(plot(eulerobj,
                                   quantities = TRUE, 
                                   labels = c("WT", "MMR-", "Same GATC", "Same Adenine"),
                                   fills = list(fill = c("#FFFF33", "#377EB8", "#278924", "#016300"),
                                   alpha = c(0.9, 0.9, 1, 1))))
}
# Look at them to see which one looks the best
plot_grid(plotlist = plot_list)
# Choose the one you like
dm_euler_plot <- plot_list[[26]]
#ggsave(dm_euler_plot,
#       filename = "/Users/carlstone/Library/CloudStorage/Box-Box/Behringer_Lab_Box_Drive/Projects/LongTermExpEvo/MethylationProject/Manuscript/Figures/venn_dm_sites.png")
#ggsave(dm_euler_plot,
#       filename = "/Users/carlstone/Library/CloudStorage/Box-Box/Behringer_Lab_Box_Drive/Projects/LongTermExpEvo/MethylationProject/Manuscript/Figures/venn_dm_sites.pdf")
dm_euler_plot <- plot(dm_overlaps,
                      fills = list(fill = c("#FFFF33", "#377EB8", "#278924", "#016300"),
                                   alpha = c(0.9, 0.9, 1, 1)))
# Recolor the intersect to be this color "#4DAF4A"

# Now import plot after some photoshopping
dm_euler_image <- cowplot::ggdraw() +
  cowplot::draw_image("/Users/carlstone/Library/CloudStorage/Box-Box/Behringer_Lab_Box_Drive/Projects/LongTermExpEvo/MethylationProject/Manuscript/Figures/venn_dm_sites_final.svg")



```

```{r figure_4_part_2}
figure_4 <- ggdraw(upset_plot) +
  draw_image("/Users/carlstone/Library/CloudStorage/Box-Box/Behringer_Lab_Box_Drive/Projects/LongTermExpEvo/MethylationProject/Manuscript/Figures/venn_dm_sites_final.svg",
             width = 0.45,
             height = 0.45,
             x = -0.06,
             y = 0.555) +
  draw_label("A",
             x = 0.015,
             y = 0.95,
             fontface = "bold") +
  draw_label("B",
             x = 0.35,
             y = 0.95,
             fontface = "bold")
ggview(figure_4,
       height = 5.5,
       width = 8, units = "in")

#ggsave(plot = figure_4,
#       filename = "/Users/carlstone/Library/CloudStorage/Box-Box/Behringer_Lab_Box_Drive/Projects/LongTermExpEvo/MethylationProject/Manuscript/Figures/new_figure_4.png",
#       height = 5.5,
#       width = 8,
#       units = "in")
#ggsave(plot = figure_4,
#       filename = "/Users/carlstone/Library/CloudStorage/Box-Box/Behringer_Lab_Box_Drive/Projects/LongTermExpEvo/MethylationProject/Manuscript/Figures/new_figure_4.pdf",
#       height = 5.5,
#       width = 8,
#       units = "in")
```

```{r print_session_info}
sessionInfo()
```

